pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,412 @@
+			return true, nil
+		}
+	}
+
+	prelimStrategy, err := pickBestStrategy(candidates)
+	if err != nil {
+		return false, err
+	}
+	if prelimStrategy != v1.OldestEmulationVersion {
+		klog.V(2).Infof("strategy %s is not recognized by CLE.", prelimStrategy)
+		return false, nil
+	}
+
+	prelimElectee := pickBestLeaderOldestEmulationVersion(candidates)
+	if prelimElectee == nil {
+		return false, nil
+	} else if lease != nil && lease.Spec.HolderIdentity != nil && prelimElectee.Name == *lease.Spec.HolderIdentity {
+		klog.V(2).Infof("Leader %s is already most optimal for lease %s %s", prelimElectee.Name, lease.Namespace, lease.Name)
+		return false, nil
+	}
+	return true, nil
+}
+
+// reconcileElectionStep steps through a step in an election.
+// A step looks at the current state of Lease and LeaseCandidates and takes one of the following action
+// - do nothing (because leader is already optimal or still waiting for an event)
+// - request ack from candidates (update LeaseCandidate PingTime)
+// - finds the most optimal candidate and elect (update the Lease object)
+// Instead of keeping a map and lock on election, the state is
+// calculated every time by looking at the lease, and set of available candidates.
+// PingTime + electionDuration > time.Now: We just asked all candidates to ack and are still waiting for response
+// PingTime + electionDuration < time.Now: Candidate has not responded within the appropriate PingTime. Continue the election.
+// RenewTime + 5 seconds > time.Now: All candidates acked in the last 5 seconds, continue the election.
+func (c *Controller) reconcileElectionStep(ctx context.Context, leaseNN types.NamespacedName) (requeue bool, err error) {
+	now := time.Now()
+
+	candidates, err := c.listAdmissableCandidates(leaseNN)
+	if err != nil {
+		return true, err
+	} else if len(candidates) == 0 {
+		return false, nil
+	}
+	klog.V(4).Infof("reconcileElectionStep %q %q, candidates: %d", leaseNN.Namespace, leaseNN.Name, len(candidates))
+
+	// Check if an election is really needed by looking at the current lease
+	// and set of candidates
+	needElection, err := c.electionNeeded(candidates, leaseNN)
+	if !needElection || err != nil {
+		return needElection, err
+	}
+
+	fastTrackElection := false
+
+	for _, candidate := range candidates {
+		// If a candidate has a PingTime within the election duration, they have not acked
+		// and we should wait until we receive their response
+		if candidate.Spec.PingTime != nil {
+			if candidate.Spec.PingTime.Add(electionDuration).After(now) {
+				// continue waiting for the election to timeout
+				return false, nil
+			} else {
+				// election timed out without ack. Clear and start election.
+				fastTrackElection = true
+				clone := candidate.DeepCopy()
+				clone.Spec.PingTime = nil
+				_, err := c.leaseCandidateClient.LeaseCandidates(clone.Namespace).Update(ctx, clone, metav1.UpdateOptions{})
+				if err != nil {
+					return false, err
+				}
+				break
```

Comment:
this is a problem: the ping times are not the same for all the objects. Due to congestion, priority and fairness or client-side throttling they could lay second apart. Some candidates could have no time to renew, while some others are pinged earlier.

Reply:
it could be seconds, yes.

The timing here seems to be brittle. Idea how to improve this:

Consider the expiry time of the lease. Ping the candidates. Give every candidate the election duration to renew the canidate. Only continue if every ping timed out individually. Do not re-ping if all renewals or pings are after lease expiry, i.e. only ping once.